src/main/sensors/rpm_filter.c

```diff
@@ -0,0 +1,204 @@
+    numberRpmNotchFilters = 0;
+    if (config->gyro_rpm_notch_harmonics) {
+        gyroFilter = &filters[numberRpmNotchFilters++];
+        rpmNotchFilterInit(gyroFilter, config->gyro_rpm_notch_harmonics,
+                           config->gyro_rpm_notch_min, config->gyro_rpm_notch_q, gyro.targetLooptime);
+    }
+    if (config->dterm_rpm_notch_harmonics) {
+        dtermFilter = &filters[numberRpmNotchFilters++];
+        rpmNotchFilterInit(dtermFilter, config->dterm_rpm_notch_harmonics,
+                           config->dterm_rpm_notch_min, config->dterm_rpm_notch_q, pidLooptime);
+    }
+
+    for (int i = 0; i < getMotorCount(); i++) {
+        pt1FilterInit(&rpmFilters[i], pt1FilterGain(RPM_MOTOR_FILTER_CUTOFF, gyro.targetLooptime));
+    }
+
+    const float secondsPerMinute = 60.0f;
+    const float erpmPerLsb = 100.0f;
+    erpmToHz = erpmPerLsb / secondsPerMinute  / (motorConfig()->motorPoleCount / 2.0f);
+
+    const float minUpdateT = 0.001f;
+    const float loopIterationsPerUpdate = minUpdateT / (pidLooptime * 1e-6f);
+    numberFilters = getMotorCount() * (filters[0].harmonics + filters[1].harmonics);
+    const float filtersPerLoopIteration = numberFilters / loopIterationsPerUpdate;
+    filterUpdatesPerIteration = rintf(filtersPerLoopIteration + 0.49f);
+}
+
+static float applyFilter(rpmNotchFilter_t* filter, int axis, float value)
+{
+    if (filter == NULL) {
+        return value;
+    }
+    for (int motor = 0; motor < getMotorCount(); motor++) {
+        for (int i = 0; i < filter->harmonics; i++) {
+            value = biquadFilterApplyDF1(&filter->notch[axis][motor][i], value);
+        }
+    }
+    return value;
+}
+
+float rpmFilterGyro(int axis, float value)
+{
+    return applyFilter(gyroFilter, axis, value);
+}
+
+float rpmFilterDterm(int axis, float value)
+{
+    return applyFilter(dtermFilter, axis, value);
+}
+
+static float motorFrequency[MAX_SUPPORTED_MOTORS];
+
+void rpmFilterUpdate()
+{
+    if (gyroFilter == NULL && dtermFilter == NULL) {
+        return;
+    }
+
+    for (int motor = 0; motor < getMotorCount(); motor++) {
+        filteredMotorErpm[motor] = pt1FilterApply(&rpmFilters[motor], getDshotTelemetry(motor));
+        DEBUG_SET(DEBUG_RPM_FILTER, motor, motorFrequency[motor]);
+    }
+
+    static uint8_t motor;
+    static uint8_t harmonic;
+    static uint8_t filter;
+    static rpmNotchFilter_t* currentFilter = &filters[0];
+
+    for (int i = 0; i < filterUpdatesPerIteration; i++) {
+
+        float frequency = (harmonic + 1) * motorFrequency[motor];
+        const float deactivateFreq = 1000.0f;
+        if (frequency < currentFilter->minHz) {
+            if (frequency < 0.5f * currentFilter->minHz) {
+                frequency = deactivateFreq;
+            } else {
+                frequency = currentFilter->minHz;
+            }
+        }
+
+        biquadFilter_t* template = &currentFilter->notch[0][motor][harmonic];
+        /* DEBUG_SET(DEBUG_RPM_FILTER, 0, harmonic); */
```

Comment:
Please don't leave commented out code in, unless there is a comment defining when this has to be re-enabled / removed. Reason: Commented out code tends to rot quickly, and it is a burden for users working on the same code, and temporarily commenting out code later on.

Reply:
ok

Reply:
> Please don't leave commented out code in, unless there is a comment defining when this has to be re-enabled / removed. Reason: Commented out code tends to rot quickly, and it is a burden for users working on the same code, and temporarily commenting out code later on.

You started preaching clean code principles!

Reply:
@stawiski: Preaching is for communicating untestable beliefs (aka. religions) - I tend to encourage / enforce patterns that are based on practical experience of what works and what doesn't.